libp2p/ENR.cpp

```diff
@@ -0,0 +1,149 @@
+// Aleth: Ethereum C++ client, tools and libraries.
+// Copyright 2019 Aleth Authors.
+// Licensed under the GNU General Public License, Version 3.
+
+#include "ENR.h"
+#include <libdevcore/SHA3.h>
+
+namespace dev
+{
+namespace p2p
+{
+namespace
+{
+constexpr char c_keyID[] = "id";
+constexpr char c_keySec256k1[] = "secp256k1";
+constexpr char c_keyIP[] = "ip";
+constexpr char c_keyTCP[] = "tcp";
+constexpr char c_keyUDP[] = "udp";
+constexpr char c_IDV4[] = "v4";
+constexpr size_t c_ENRMaxSize = 300;
+
+
+// Address can be either boost::asio::ip::address_v4 or boost::asio::ip::address_v6
+template <class Address>
+bytes addressToBytes(Address const& _address)
+{
+    auto const addressBytes = _address.to_bytes();
+    return bytes(addressBytes.begin(), addressBytes.end());
+}
+}  // namespace
+
+ENR::ENR(RLP _rlp, VerifyFunction _verifyFunction)
+{
+    if (_rlp.data().size() > c_ENRMaxSize)
+        BOOST_THROW_EXCEPTION(ENRIsTooBig());
+
+    m_signature = _rlp[0].toBytes(RLP::VeryStrict);
+
+    m_seq = _rlp[1].toInt<uint64_t>(RLP::VeryStrict);
+
+    // read key-values into vector first, to check the order
+    std::vector<std::pair<std::string const, bytes>> keyValues;
+    for (size_t i = 2; i < _rlp.itemCount(); i+= 2)
```

Comment:
`count - 1` looks dangerous for unsigned type (well in this case it shouldn't underflow though)

Also `_rlp[i + 1]` in the body loop should throw for out of bounds and not lead to UB.

Reply:
I think _rlp[i + 1] would only throw if it happened to point to protected memory (unless you have something like[ full page heap enabled](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/gflags-and-pageheap)).

Reply:
I think you confuse it with arrays/vectors.

Actually I was not right - `RLP::operator[]` will return empty `RLP` for out of bounds - I forget this all the time. So in this case it will end up with `key => ""` if last value is absent, which is not ideal, but it's not an UB, I think I'll merge it like this for now.

Reply:
I've added some comments to `RLP::operator[]` which is not very straight-forward https://github.com/ethereum/aleth/pull/5555/commits/a791c5b2427606da9bde536b86c74b36f4df5d29